### ISSUE TITLE
Can't load large feature service into qgis 3.10

### ISSUE THREAD

**Issue by soneil1** [qgis/QGIS#34970]
i get this error when trying to load a large feature service layer into qgis. Another layer that's smaller in size within the same service loads fine.
**Layer is not valid: The layer crs='EPSG:26918' url='https://gisservices.its.ny.gov/arcgis/rest/services/BuildingFootprints/FeatureServer/2' is not a valid layer and can not be added to the map. Reason: AFSProvider getObjectIds failed: Parsing error - illegal value**

**Comment by gioman** [comment]
@soneil1 does it loads fine on other GIS software?

**Comment by soneil1** [comment]
Yes, it loads in ArcMap

**Comment by gioman** [comment]
There is also timeout error in QGIS logs

```
2020-03-10T14:37:15     WARNING    Network request https://gisservices.its.ny.gov/arcgis/rest/services/BuildingFootprints/FeatureServer/2/query?f=json&where=1%3D1&returnIdsOnly=true timed out
2020-03-10T14:44:17     WARNING    Host not found
```

that could be very well the root cause of the issue...

**Comment by haubourg** [comment]
@soneil1 can you try to raise your timeout delay in general settings / network and see if it helps?

**Comment by soneil1** [comment]
I have the timeout set to 180000

**Comment by gioman** [comment]
> I have the timeout set to 180000

then maybe is just a super slow endpoint.

**Comment by Saijin-Naib** [comment]
![image](https://user-images.githubusercontent.com/19295950/76324408-56ed7e00-62bc-11ea-8a6b-4c6c4c00ef3c.png)
`WARNING    Network request https://gisservices.its.ny.gov/arcgis/rest/services/BuildingFootprints/FeatureServer//2/query?f=json&where=1%3D1&returnIdsOnly=true timed out`
Interesting... Tested locally here on 3.12 and received the same Message as above, but only after an attempted connection took in excess of 30s.

Layer 1 (the availability footprint) loads just fine (near-instant). It does seem to corroborate his report.

@soneil1, as an aside, there is decent building footprint data that Microsoft provided to OpenStreetMap, and the NYS GIS Clearinghouse has that available for download as well.

**Comment by soneil1** [comment]
Yes, the availability layer works fine but it's a smaller layer. The footprints layer has 7 million records in it. 

@Saijin-Naib we do have the microsoft layer, we're actually having a separate publishing issue with that one right now.

**Comment by soneil1** [comment]
Do you know if there's anything I can change on the server side?

**Comment by gioman** [comment]
> Do you know if there's anything I can change on the server side?

@soneil1 use WFS? ;)

**Comment by soneil1** [comment]
@gioman we do have wfs enabled!

**Comment by Saijin-Naib** [comment]
> Yes, the availability layer works fine but it's a smaller layer. The footprints layer has 7 million records in it.
> 
> @Saijin-Naib we do have the microsoft layer, we're actually having a separate publishing issue with that one right now.

Nice to e-meet you, as a consumer/resident ;)

It is quite odd that Pro will load/connect to the footprints instantly and QGIS chokes...

**Comment by gioman** [comment]
> @gioman we do have wfs enabled!

@soneil1 URL?

**Comment by soneil1** [comment]
@Saijin-Naib I agree

**Comment by soneil1** [comment]
@gioman  http://gisservices.its.ny.gov/arcgis/services/BuildingFootprints/MapServer/WFSServer?request=GetCapabilities&service=WFS

**Comment by gioman** [comment]
> @gioman http://gisservices.its.ny.gov/arcgis/services/BuildingFootprints/MapServer/WFSServer?request=GetCapabilities&service=WFS

@soneil1 works but it just loads 1000 features, have you set some limit server side?

**Comment by soneil1** [comment]
@gioman yes it's set at 1000 for max number of features returned by service. we might need to up that if the feature service won't load in qgis.

**Comment by gioman** [comment]
> @gioman yes it's set at 1000 for max number of features returned by service. we might need to up that if the feature service won't load in qgis.

@soneil1 
if we want to understand where the issue here we need to test both services from the same server with the same full dataset.

**Comment by soneil1** [comment]
@gioman in our parameter properties our max number of records returned is set to 50,000. I change the wfs max number to 50,000 as well.

**Comment by elpaso** [comment]
Be aware of this upstream bug too: https://bugreports.qt.io/browse/QTBUG-76821

**Comment by gioman** [comment]
> @gioman in our parameter properties our max number of records returned is set to 50,000. I change the wfs max number to 50,000 as well.

@soneil1 don't know.... but the service seems dead slow... in fact after adding the layer QGIS never loads any feature on map. And no, is not QGIS or my connection. I tested a different WFS server (based on Mapserver) with a layer with 6+ million features and it works ok.

**Comment by nyalldawson** [comment]
>  The footprints layer has 7 million records in it.

FeatureServer is a horrendously inefficient protocol for transmission of large layers like this. It's an extremely verbose text based format, so you'll be waiting for hours for this amount of data to transfer.

A potential workaround is to restrict the layer's contents to the current bounding box of the canvas by clicking this option:
![image](https://user-images.githubusercontent.com/1829991/76365601-64b20a80-6373-11ea-850b-302e5abea756.png)

That will limit the layer to only retrieving the features visible in the current bounding box.

(I think this report should be left open, because there is **some** work we could do to defer the initial afs provider initialisation into the feature iterator, so at least it wouldn't block the whole qgis window while this download occurs)

**Comment by PeterReyne** [comment]
Hi,

I also came across this issue a month or so ago while doing some other testing with QGIS and ArcGIS Server 10.7.1

**The root cause:** ArcGIS Server doesn't apply the configured query feature limit to requests that have "returnIdsOnly=true" param set and the response is quite slow for large datasets.

**The potential fix:** QGIS could use "resultRecordCount" and "resultRecordOffset" params to limit the response to a certain number of records (e.g. 1000) like below.

https://gisservices.its.ny.gov/arcgis/rest/services/BuildingFootprints/FeatureServer/2/query?f=json&where=1%3D1&returnIdsOnly=true&resultRecordCount=1000&resultOffset=0

an extension:
QGIS could set the Query "resultRecordCount" value using the "maxRecordCount" value from this request
https://gisservices.its.ny.gov/arcgis/rest/services/BuildingFootprints/FeatureServer/2?f=pjson

**Comment by Pedro-Murteira** [comment]
@soneil1 Still true on recent releases?
